src/mscorlib/shared/System/IO/Path.cs

```diff
@@ -23,6 +23,7 @@ public static partial class Path
         // For generating random file names
         // 8 random bytes provides 12 chars in our encoding for the 8.3 name.
```

Comment:
This makes again wonder whether this is a good API to add and evolve. 8.3 is legacy Windowsism.

@JeremyKuhne Do we know that this is a good API to have from a security and robustness point of view? Why is 8.3 randomness good for everybody? Should we rather have an API that takes random bytes (that you generate using random number generator of your choice and strength) and produces a valid filename out of it?

Reply:
@morganbr @GrabYourPitchforks

Reply:
> FAT is not.

The current FAT has support for longer file names. I doubt there is anybody who still cares about the very old 8.3 FAT.

Reply:
> The current FAT has support for longer file names. I doubt there is anybody who still cares about the very old 8.3 FAT.

Windows implementation does, sure, but there are plenty of others. Additionally this avoids generating short filenames.

Reply:
Why the thumbs down?

Reply:
For the record- short file names create crazy hard bugs. I'm following one right now that looks like it might be related to how intermediate I/O filters may be causing unexpected 8.3 matches in rare cases.  I explicitly removed 8.3 matching in Core 2.1 to make matching more dependable. Even though we aren't impacted by this now, native code and other apps are.

Reply:
Because of I disagree with what you have said.

Reply:
I do not think we should be designing for Windows legacy short file names at all.

We should be designing as it the Windows legacy short file names did not exist. Once the design is done, we can check how well it works for Windows legacy short file names just so that we know where we are. However, I do not see a problem with new things not working well with Windows legacy short filenames.

Reply:
> We should be designing as it the Windows legacy short file names did not exist. 

It is a shipped API, and it has value. An 8.3 file name with these characters is the most broadly portable and has the least possible side-effects. Giving up those design points does not justify what we would potentially gain as there are scenarios where new behavior would not work or work as well.

The existence of this API does not preclude adding ones with differing behavior. Not having it creates unnecessary work and risk for changing existing code to minimize allocations. I'm not against adding additional APIs with differing behavior, but not at the cost of losing this one.

Reply:
> It is a shipped API, and it has value. 

This does not imply that non-allocating version of the API has value too.

We started with the same position when adding the Span overloads for the existing String APIs (https://github.com/dotnet/corefx/issues/21395). It turned out that that non-allocating versions of number of String APIs did not make sense after we looked into whether and how would anybody use them. Number of APIs from the original proposal were dropped, and number had design changed.

It proved to be valuable to ask questions like: Who is this non-allocating API for? Is there an real-world existing code that will get signficant benefit from using the non-allocating API?

I looked around, but I have not found anything:

msbuild is a very large path manipulation app. They use GetRandomFileName in a few tests: https://github.com/Microsoft/msbuild/search?utf8=%E2%9C%93&q=GetRandomFileName&type= . There is no point in minimizing allocations in tests.

From searching github https://github.com/search?q=GetRandomFileName++language%3Acsharp&type=Code, it seems that the most popular use of Path.GetRandomFileName has nothing to do with paths. People use it as a convenient way to generate a random string, typically for testing. And they often do not want the dot in the name and run `Replace(".", string.Empty)` to get rid of it. There are some uses of where it is used as intended for paths, but none of them look performance critical.

I have also checked how often Microsoft internal apps call GetRandomFileName in Azure. azureprofiler has zero hits for it.

Reply:
The basic canonical use of this API is: `Path.Combine(Path.GetTempPath(), Path.GetRandomFileName())`

What is no-allocation code that you expect folks will write instead of this?

Reply:
> The basic canonical use of this API is: `Path.Combine(Path.GetTempPath(), Path.GetRandomFileName())`
> What is no-allocation code that you expect folks will write instead of this?

Less-allocation for the canonical case currently. We have a `TryCombine` that would eliminate an extra allocation there (two if you cache the temp path or we add the other api). Additionally, we'll eventually add `File` and `Directory` APIs that take Span that will benefit even more in the future. Finding an unused name, copying files (say, for backing up or shadow caching), etc. will get some benefit now, more later.

Outside of that, as you've noticed, this is a common "random string" generator.

Reply:
I understand that it is some combination of the APIs that you have mentioned, but I would love to see the actual code one has to write.

Reply:
> I understand that it is some combination of the APIs that you have mentioned, but I would love to see the actual code one has to write.

I could write you an example of `public FileStream GetTemporaryFile()` and how that would evolve as additional APIs are added.

Reply:
I think rewrite of this https://github.com/cdorst/App.Prototype/blob/master/ConsoleApp6/Functions/TempDirectoryGetter.cs (randomly picked example) using this new API would be enough to demonstrate it.

The function would still allocate the final string, but without the intermediate allocations.

Reply:
That one would be something like:

``` C#
public static class TempDirectoryGetter
{
    private const string repo = nameof(repo);

    // As in other discussions this can be safely cached and as such we don't
    // strictly need a Try* for this.
    private static string s_tempPath = Path.GetTempPath();

    public unsafe static string GetTempDirectory(string subdirectory = repo)
    {
        Span<char> buffer = stackalloc char[12];
        Path.TryGetRandomFileName(buffer, out _);
        return Path.Join(s_tempPath, subdirectory, buffer);
    }
}
```

Nice to lose an allocation that we couldn't otherwise. The example I mentioned where you would potentially loop looking for free names is a bit more interesting, but more complicated (and would benefit further from `Exists(Span<char>)`).

Reply:
How would you do it without hardcoding `12`? If any reasonable use of this API has to have `12` hardcoded and ignores return value and charsWritten, the API signature can be much simpler.

Reply:
I would _prefer_ that the API be `Span<char> TryGetRandomFileName(Span<char> buffer)` and have it return an empty span if it can't fill it.  That isn't like the other Try methods however.

``` C#
public static class TempDirectoryGetter
{
    private const string repo = nameof(repo);

    // As in other discussions this can be safely cached and as such we don't
    // strictly need a Try* for this.
    private static string s_tempPath = Path.GetTempPath();

    public unsafe static string GetTempDirectory(string subdirectory = repo)
    {
        Span<char> buffer = stackalloc char[12];
        return Path.Join(s_tempPath, subdirectory, Path.TryGetRandomFileName(buffer));
    }
}
```